File: app/src/main/kotlin/tech/mmarca/openvitals/features/activity/ActivityHeartRateChartCard.kt
package tech.mmarca.openvitals.features.activity

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import kotlin.math.roundToInt
import kotlin.math.roundToLong
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import androidx.health.connect.client.records.ExerciseSegment
import tech.mmarca.openvitals.domain.model.BleHeartRateSample
import tech.mmarca.openvitals.domain.model.BleRecordingSampleBuffer
import tech.mmarca.openvitals.domain.model.ExerciseSegmentData
import tech.mmarca.openvitals.domain.model.HeartRateSample
import tech.mmarca.openvitals.ui.components.ChartXAxisWithYAxis
import tech.mmarca.openvitals.ui.components.ChartZoom
import tech.mmarca.openvitals.ui.components.MetricLinePlot
import tech.mmarca.openvitals.ui.components.MetricLinePlotPoint
import tech.mmarca.openvitals.ui.components.OpenVitalsCard
import tech.mmarca.openvitals.ui.components.SessionAxis
import tech.mmarca.openvitals.ui.components.SessionAxisLabels
import tech.mmarca.openvitals.ui.components.SessionPause
import tech.mmarca.openvitals.ui.components.formatElapsedChartLabel
import tech.mmarca.openvitals.ui.theme.HeartColor

@Composable
internal fun ActivityHeartRateChartCard(
    samples: List<HeartRateSample>,
    sessionStart: Instant,
    sessionEnd: Instant,
    unitFormatter: UnitFormatter,
    modifier: Modifier = Modifier,
    pauses: List<SessionPause> = emptyList(),
) {
    if (samples.isEmpty()) return

    val sorted = samples.sortedBy { it.time }
    val minBpm = sorted.minOf { it.beatsPerMinute }
    val maxBpm = sorted.maxOf { it.beatsPerMinute }
    val avgBpm = sorted.map { it.beatsPerMinute }.average().roundToInt()
    val paddedMin = (minBpm - 5L).coerceAtLeast(30L)
    val paddedMax = maxBpm + 5L
    // The axis counts only moving time: a pause is not part of the ride, so it
    // gets none of the chart. Heart rate keeps sampling through a pause (a strap
    // does not stop because the ride did); those samples stack at the moment the
    // pause began rather than stretching it back open.
    val axis = remember(sessionStart, sessionEnd, pauses) {
        SessionAxis(start = sessionStart, end = sessionEnd, pauses = pauses)
    }
    val chartHeight = 180.dp
    val zone = ZoneId.systemDefault()
    val timeFormatter = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)
    // Built once, identity-stable across pinch frames, so the plot's geometry
    // cache holds.
    val chartPoints = remember(sorted, axis) {
        sorted.map { sample ->
            MetricLinePlotPoint(
                xFraction = axis.fractionOf(sample.time),
                value = sample.beatsPerMinute.toDouble(),
            )
        }
    }

    OpenVitalsCard(modifier = modifier) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = stringResource(R.string.activity_recording_live_heart_rate),
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(16.dp),
            ) {
                ActivityHeartRateStat(
                    label = stringResource(R.string.summary_average),
                    value = unitFormatter.heartRate(avgBpm.toLong()).text,
                    modifier = Modifier.weight(1f),
                )
                ActivityHeartRateStat(
                    label = stringResource(R.string.summary_range),
                    value = "${unitFormatter.heartRate(minBpm).text}-${unitFormatter.heartRate(maxBpm).text}",
                    modifier = Modifier.weight(1f),
                )
                ActivityHeartRateStat(
                    label = stringResource(R.string.summary_samples),
                    value = unitFormatter.count(sorted.size),
                    modifier = Modifier.weight(1f),
                )
            }
            // The plot and its elapsed row share the one viewport, so the labels
            // always describe the slice of the session actually on show.
            ChartZoom(sessionStart, sessionEnd, samples) { zoom ->
                Column {
                    MetricLinePlot(
                        points = chartPoints,
                        minValue = paddedMin.toDouble(),
                        maxValue = paddedMax.toDouble(),
                        accentColor = HeartColor,
                        chartHeight = chartHeight,
                        valueFormatter = { unitFormatter.heartRate(it.roundToLong()).text },
                        pointRadius = if (sorted.size <= 120) 2.dp else 0.dp,
                        lineStrokeWidth = 2.dp,
                        drawPoints = sorted.size <= 120,
                        viewport = zoom.viewport,
                        multiTouch = zoom.multiTouch,
                        // Moving elapsed, matching the labels directly under it:
                        // reporting wall-clock here would have the scrubber
                        // disagree with its own axis.
                        scrubLabel = { point ->
                            unitFormatter.heartRate(point.value.roundToLong()).text to
                                formatElapsedChartLabel(axis.elapsedAt(point.xFraction))
                        },
                    )
                    Spacer(Modifier.height(4.dp))
                    ChartXAxisWithYAxis {
                        SessionAxisLabels(axis = axis, viewport = zoom.viewport)
                    }
                }
            }
            Text(
                text = stringResource(
                    R.string.summary_recorded,
                    timeFormatter.format(sorted.first().time.atZone(zone)),
                    timeFormatter.format(sorted.last().time.atZone(zone)),
                ),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }
}

@Composable
private fun ActivityHeartRateStat(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier) {
        Text(
            text = value,
            style = MaterialTheme.typography.titleMedium,
            color = HeartColor,
        )
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

internal fun BleRecordingSampleBuffer.toHeartRateSamples(): List<HeartRateSample> =
    heartRateSamples.map { sample ->
        HeartRateSample(
            time = sample.time,
            beatsPerMinute = sample.beatsPerMinute,
            source = "sensor",
        )
    }

internal fun List<BleHeartRateSample>.toHeartRateSamples(): List<HeartRateSample> =
    map { sample ->
        HeartRateSample(
            time = sample.time,
            beatsPerMinute = sample.beatsPerMinute,
            source = "sensor",
        )
    }

/** The pause segments of a workout, as the pauses the session axis skips. */
internal fun List<ExerciseSegmentData>.toSessionPauses(): List<SessionPause> =
    filter { it.segmentType == ExerciseSegment.EXERCISE_SEGMENT_TYPE_PAUSE }
        .map { SessionPause(start = it.startTime, end = it.endTime) }

File: lib/features/activity/activity_heart_rate_chart_card.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../../core/presentation/unit_formatter.dart';
import '../../domain/model/heart_models.dart';
import '../../l10n/app_localizations.dart';
import '../../ui/charts/chart_axis.dart';
import '../../ui/charts/metric_line_plot.dart';
import '../../ui/components/ov_card.dart';
import '../../ui/theme/app_colors.dart';
import '../manualentry/activity/recording/activity_recording_dashboard.dart'
    show formatRecordingElapsed;

/// Port of the Kotlin `ActivityHeartRateChartCard`: the recorded heart-rate
/// trace on a real time axis keyed to the session range, with the avg / range /
/// sample-count stat row above it and elapsed-time labels below.
class ActivityHeartRateChartCard extends StatelessWidget {
  const ActivityHeartRateChartCard({
    super.key,
    required this.samples,
    required this.sessionStart,
    required this.sessionEnd,
    required this.unitFormatter,
  });

  final List<HeartRateSample> samples;
  final DateTime sessionStart;
  final DateTime sessionEnd;
  final UnitFormatter unitFormatter;

  /// Kotlin `chartHeight`.
  static const double _chartHeight = 180;

  @override
  Widget build(BuildContext context) {
    if (samples.isEmpty) return const SizedBox.shrink();

    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);
    final locale = Localizations.localeOf(context).toString();

    final sorted = [...samples]..sort((a, b) => a.time.compareTo(b.time));
    final bpms = [for (final sample in sorted) sample.beatsPerMinute];
    final minBpm = bpms.reduce(math.min);
    final maxBpm = bpms.reduce(math.max);
    final avgBpm = (bpms.reduce((a, b) => a + b) / bpms.length).round();
    // Kotlin pads the Y range by 5 bpm, flooring the minimum at 30.
    final paddedMin = math.max(minBpm - 5, 30);
    final paddedMax = maxBpm + 5;
    final sessionDurationMillis =
        math.max(sessionEnd.difference(sessionStart).inMilliseconds, 1);
    final drawPoints = sorted.length <= 120;
    final timeFormatter = DateFormat.jm(locale);

    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 12,
          children: [
            Text(
              l10n.activityRecordingLiveHeartRate,
              style: theme.textTheme.titleMedium
                  ?.copyWith(fontWeight: FontWeight.bold),
            ),
            Row(
              spacing: 16,
              children: [
                Expanded(
                  child: _ActivityHeartRateStat(
                    label: l10n.summaryAverage,
                    value: unitFormatter.heartRate(avgBpm).text,
                  ),
                ),
                Expanded(
                  child: _ActivityHeartRateStat(
                    label: l10n.summaryRange,
                    value: '${unitFormatter.heartRate(minBpm).text}-'
                        '${unitFormatter.heartRate(maxBpm).text}',
                  ),
                ),
                Expanded(
                  child: _ActivityHeartRateStat(
                    label: l10n.summarySamples,
                    value: unitFormatter.count(sorted.length),
                  ),
                ),
              ],
            ),
            MetricLinePlot(
              points: [
                for (final sample in sorted)
                  MetricLinePlotPoint(
                    xFraction: sample.time
                            .difference(sessionStart)
                            .inMilliseconds
                            .clamp(0, sessionDurationMillis) /
                        sessionDurationMillis,
                    value: sample.beatsPerMinute.toDouble(),
                  ),
              ],
              minValue: paddedMin.toDouble(),
              maxValue: paddedMax.toDouble(),
              accentColor: AppColors.heart,
              chartHeight: _chartHeight,
              valueFormatter: (value) =>
                  unitFormatter.heartRate(value.round()).text,
              pointRadius: drawPoints ? 2 : 0,
              lineStrokeWidth: 2,
              drawPoints: drawPoints,
            ),
            const SizedBox(height: 4),
            ChartXAxisWithYAxis(
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  for (final label in sessionElapsedLabels(
                      Duration(milliseconds: sessionDurationMillis)))
                    Text(
                      label,
                      style: theme.textTheme.labelSmall?.copyWith(
                        color: theme.colorScheme.onSurfaceVariant,
                      ),
                    ),
                ],
              ),
            ),
            Text(
              l10n.summaryRecorded(
                timeFormatter.format(sorted.first.time.toLocal()),
                timeFormatter.format(sorted.last.time.toLocal()),
              ),
              style: theme.textTheme.bodySmall?.copyWith(
                color: theme.colorScheme.onSurfaceVariant,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

/// Kotlin `ActivityHeartRateStat`.
class _ActivityHeartRateStat extends StatelessWidget {
  const _ActivityHeartRateStat({required this.label, required this.value});

  final String label;
  final String value;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          value,
          style: theme.textTheme.titleMedium?.copyWith(color: AppColors.heart),
        ),
        Text(
          label,
          style: theme.textTheme.labelSmall?.copyWith(
            color: theme.colorScheme.onSurfaceVariant,
          ),
        ),
      ],
    );
  }
}

/// Kotlin `sessionElapsedLabels`: X-axis labels at 0, ¼, ½, ¾ and the full
/// session duration.
List<String> sessionElapsedLabels(Duration duration) => [
      formatRecordingElapsed(Duration.zero),
      formatRecordingElapsed(duration ~/ 4),
      formatRecordingElapsed(duration ~/ 2),
      formatRecordingElapsed(duration * 3 ~/ 4),
      formatRecordingElapsed(duration),
    ];
